Code review for HashStore Beta release 1.0.0

doulikecookiedough opened the pull request:
Can you please provide feedback regarding the [public API methods](https://github.com/NCEAS/metacat/blob/feature-1436-storage-and-indexing/docs/user/metacat/source/storage-subsystem.rst), existing tests (tests/test_hashstore.py) and any concerns regarding the existing behaviour? Changes/requests that will not be quick to resolve will be organized into issues & discussion items.

doulikecookiedough commented on src/hashstore/hashstore.py:
```diff
@@ -143,7 +148,7 @@ def _add_object(self, data, algorithm, checksum):
         address = self.objects.put(data, algorithm=algorithm, checksum=checksum)
         if address.is_duplicate:
-            return None
+            raise FileExistsError("Object already exists.")
```
Updated info message as per your suggestion in commit [3036ab3](https://github.com/DataONEorg/hashstore/pull/14/commits/3036ab3b72c65df34e9ee3bd43338739d92cc610)

doulikecookiedough commented on src/hashstore/hashstore.py:
```diff
@@ -206,3 +211,170 @@ def _rel_path(self, hash):
+        "HashAddress", ["id", "relpath", "abspath", "is_duplicate", "hex_digests"]
+    )
+):
+    """File address containing file's path on disk and it's content hash ID.
```
Fixed typo, thank you!

doulikecookiedough commented on tests/test_hashstore.py:
Added additional test for testing an input_stream in [0d44513](https://github.com/DataONEorg/hashstore/pull/14/commits/0d44513e1cf16bfcd86b9ce406aaa65740ea3f05)

doulikecookiedough commented on src/hashstore/hashstore.py:
line 24:
- time_out - updated to time_out_sec

line 26/139:
- I agree with the inconsistencies you mentioned, so I added a new method "\_clean\_algorithm" that replaces "-" with a "_" when more than 3 digits are counted in the algorithm string - as well as a new test. This method is applied in the affected lines of code where an algorithm is in use. For now... I believe it should suffice, but it may need updating in the future. 
- For context, in order to create a new hash object with the hashlib library for algorithms that have more than 3 digits (like "algo#-###") - hashlib requires the desired algorithm string with underscores. This is why I decided to format the `supported_algorithms` list as it is currently (which are all values that are hashlib-compliant).

line 80:
- Updated to include passed value of `algorithm`

doulikecookiedough commented on src/hashstore/hashstore.py:
```diff
@@ -206,3 +211,170 @@ def _rel_path(self, hash):
         return "/".join(chunks)
+
+
+class HashFSExt(HashFS):
```
Added new tests to check the override methods

doulikecookiedough commented:
@artntek Thank you for reviewing my pull request! I believe I addressed all your comments - please let me know if you have any further feedback.

taojing2002 commented on src/hashstore/hashstore.py:
line 107 - What will happen if the object associated with the pid doesn't exist when a client call the retrieve_object method? Should it raise a not-found exception?

line 116 -  Use above comment for the retrieve_sysmeta method

line 160 - It has a duplicated hard code `sysmeta` to line 50. May we have a class variable in the HashFSExt class to store its root path? Also, you can have a method to get the root path. So you can eliminate those duplicated hard code.

line 187 - I believe the purpose of this line is to delete the new created system metadata file if some bad things happen.  However, if the exception happens before the new system metadata file is created (e.g. failed to create the tmp file), the code will actually delete the original system metadata and we will lose the information. Also, why do you use the api call to delete the system metadata file? It has some overhead. 

line 276 - The name of the method `copy` confuses me a little bit. I thought it will just copy a file from a place to another one. However, its function is more than that - it will save bytes into a temporary file and move the temp file to a permanent place.

line 339 - It will be better if the variable of default_algo_list is a class variable. 

line 340 - It will be better if the variable of other_algo_list is a class varaible.

doulikecookiedough commented:
Thank you @taojing2002 for your feedback!

**RE: Line 107/116**
- Updated both methods to raise ValueError if sysmeta does not exist (`f"No sysmeta found for pid: {pid}"`)

**RE: Line 160**
- Replaced the hardcoded sysmeta store reference and added new method to get the root path of the store (`_get_store_path()`)

**RE: Line 276**
- Renamed `_copy` method to `_move_and_get_checksums` for clarity

**RE: Line 339/340**
- Changed `default_algo_list` and `other_algo_list` to class variables

**RE: Line 187+**
- I removed the API call to `delete_sysmeta` (to reduce overhead) - I missed this - thank you for catching that.
- I also refactored the `set_sysmeta` method by adding a try/except block when attempting to rename the sysmeta (when preparing for an update action).
   - By doing so, my intention is to ensure that the `sysmeta.tmp` file exists before proceeding during an update
   - This will prevent the accidental deletion of the original sysmeta file when rolling back the process - i.e. if an exception occurs while writing the new sysmeta, HashStore will confirm the existence of the original sysmeta file (`sysmeta.tmp`), then delete the new sysmeta file (if it exists) before reverting (renaming) `sysmeta.tmp` back to `sysmeta`
- How do you feel about this process with this change?

doulikecookiedough commented:
Thank you again @taojing2002 and @artntek for your feedback. After our latest discussions regarding HashStore, there are further changes required that is outside the scope for this pull request review. I am going to merge this branch into develop, and then create new issues that will address the new requirements and any pending questions posed here (ex. set_sysmeta flow).
